shopfloor_base/utils.py

```diff
@@ -33,6 +34,31 @@ def wrapped(*args, **kwargs):
     return _ensure_model
 
 
+def catch_errors(response_error):
+    """Decorator for service endpoints to gracefully handle some exceptions.
+
+    The decorator is passed a function with the same arguments than the function
+    handlling the endpoint, plus a keyword argument called `message`.
+    """
+
+    def _catch_errors(fn):
+        @wraps(fn)
+        def wrapper(*args, **kwargs):
+            try:
+                res = fn(*args, **kwargs)
+            except ValidationError as ex:
+                message = {
+                    "message_type": "error",
+                    "body": ex.name,
+                }
+                return response_error(*args, **kwargs, message=message)
+            return res
+
+        return wrapper
+
+    return _catch_errors
```

Comment:
it would be nice to have a way to rollback the transaction in case of an error:
```suggestion
    def _catch_errors(fn):
        @wraps(fn)
        def wrapper(*args, **kwargs, rollback=False):
            savepoint = self._actions_for("savepoint").new()
            try:
                res = fn(*args, **kwargs)
            except ValidationError as ex:
                if rollback:
                    savepoint.rollback()
                else:
                    savepoint.release()
                message = {
                    "message_type": "error",
                    "body": ex.name,
                }
                return response_error(*args, **kwargs, message=message)
            savepoint.release()
            return res

        return wrapper

    return _catch_errors
```
or smth like that.
In this implementation, I'm also checking that response isn't an error.
https://github.com/OCA/wms/blob/14.0/shopfloor_single_product_transfer/services/single_product_transfer.py#L20

Reply:
Thank you, yes it is needed, added a fixup